src/test/ui/suggestions/lifetimes/missing-lifetimes-in-signature.stderr

```diff
@@ -0,0 +1,115 @@
+LL | |     }
+   | |_____^ ...but this borrow...
+   |
+note: ...can't outlive the anonymous lifetime #1 defined on the function body at 15:1
+  --> $DIR/missing-lifetimes-in-signature.rs:15:1
+   |
+LL | / fn foo<G, T>(g: G, dest: &mut T) -> impl FnOnce()
+LL | | where
+LL | |     G: Get<T>
+LL | | {
+...  |
+LL | |     }
+LL | | }
+   | |_^
+help: you can add a bound to the returned `impl Trait` to make it last less than `'static` and match the anonymous lifetime #1 defined on the function body at 15:1
+   |
+LL | fn foo<G, T>(g: G, dest: &mut T) -> impl FnOnce() + '_
+   |                                                   ^^^^
+
+error[E0311]: the parameter type `G` may not live long enough
+  --> $DIR/missing-lifetimes-in-signature.rs:25:37
+   |
+LL | fn bar<G, T>(g: G, dest: &mut T) -> impl FnOnce() + '_
+   |                                     ^^^^^^^^^^^^^^^^^^
+   |
+note: the parameter type `G` must be valid for the anonymous lifetime #1 defined on the function body at 25:1...
+  --> $DIR/missing-lifetimes-in-signature.rs:25:1
+   |
+LL | / fn bar<G, T>(g: G, dest: &mut T) -> impl FnOnce() + '_
+LL | |
+LL | | where
+LL | |     G: Get<T>
+...  |
+LL | |     }
+LL | | }
+   | |_^
+note: ...so that the type `[closure@$DIR/missing-lifetimes-in-signature.rs:30:5: 32:6 g:G, dest:&mut T]` will meet its required lifetime bounds
+  --> $DIR/missing-lifetimes-in-signature.rs:25:37
+   |
+LL | fn bar<G, T>(g: G, dest: &mut T) -> impl FnOnce() + '_
+   |                                     ^^^^^^^^^^^^^^^^^^
+help: consider introducing an explicit lifetime bound to unify the type parameter and the output
+   |
+LL | fn bar<'a, G: 'a, T>(g: G, dest: &mut T) -> impl FnOnce() + '_ + 'a
+   |        ^^^^^                                                   ^^^^
+
+error[E0311]: the parameter type `G` may not live long enough
+  --> $DIR/missing-lifetimes-in-signature.rs:47:45
+   |
+LL | fn qux<'a, G: 'a, T>(g: G, dest: &mut T) -> impl FnOnce() + '_
+   |                                             ^^^^^^^^^^^^^^^^^^
+   |
+note: the parameter type `G` must be valid for the anonymous lifetime #1 defined on the function body at 47:1...
+  --> $DIR/missing-lifetimes-in-signature.rs:47:1
+   |
+LL | / fn qux<'a, G: 'a, T>(g: G, dest: &mut T) -> impl FnOnce() + '_
+LL | |
+LL | | where
+LL | |     G: Get<T>
+...  |
+LL | |     }
+LL | | }
+   | |_^
+note: ...so that the type `[closure@$DIR/missing-lifetimes-in-signature.rs:52:5: 54:6 g:G, dest:&mut T]` will meet its required lifetime bounds
+  --> $DIR/missing-lifetimes-in-signature.rs:47:45
+   |
+LL | fn qux<'a, G: 'a, T>(g: G, dest: &mut T) -> impl FnOnce() + '_
+   |                                             ^^^^^^^^^^^^^^^^^^
+help: consider introducing an explicit lifetime bound to unify the type parameter and the output
+   |
+LL | fn qux<'b, 'a, G: 'b + 'a, T>(g: G, dest: &mut T) -> impl FnOnce() + '_ + 'b
+   |        ^^^     ^^^^^^^                                                  ^^^^
```

Comment:
Hmm I don't think this suggestion is going to work. I think you need something like this:

```
fn qux<'b, 'a, G: 'a + 'b, T>(g: G, dest: &'b mut T) -> impl FnOnce() + '_
```

i.e., what's going wrong here is that:

* the `'_` refers to the lifetime of the `&mut T`
* we have no connection between the type `G`  and that lifetime

I don't think that just adding `'b` as suggested here is going to solve the problem, since it doesn't address that mismatch.

Reply:
The suggestion will indeed not work in isolation, but if you look at [the output here](https://github.com/rust-lang/rust/pull/72543/files/4ff2398447a0472d05bff4612ea0d8b6b3922edd#diff-20c32fdf2e337251f1e91732f285d11dR90-R96), after applying these changes we will suggest `dest: &'b mut T`. A good thing to add would be to _remove_ the ` + '_` part, but didn't want to dig too deep into that and instead focusing on making every possible combination give errors that move you in the right direction.

Reply:
Now that I am looking at this again I think I _could_ make this suggest <code>fn qux<'a, G: 'a, T>(g: G, dest: &mut T) -> impl FnOnce() + '_ <b>+ 'a</b></code> instead. It will just take handling it on its own.

Reply:
I guess it feels like this suggestion will not really help the user know what's going on -- they have to apply it and *then* maybe they'll understand -- but maybe that's the best we can do.